jvm-runtime/ftl-runtime/common/runtime/src/main/java/xyz/block/ftl/runtime/FTLController.java

```diff
@@ -31,61 +29,41 @@
 import xyz.block.ftl.v1.VerbServiceGrpc;
 import xyz.block.ftl.v1.schema.Ref;
 
-@Singleton
-@Startup
 public class FTLController implements LeaseClient {
     private static final Logger log = Logger.getLogger(FTLController.class);
     final String moduleName;
 
     private Throwable currentError;
     private volatile ModuleContextResponse moduleContextResponse;
     private boolean waiters = false;
-    private volatile boolean closed = false;
 
     final VerbServiceGrpc.VerbServiceStub verbService;
-    final StreamObserver<ModuleContextResponse> moduleObserver = new StreamObserver<>() {
-        @Override
-        public void onNext(ModuleContextResponse moduleContextResponse) {
-            synchronized (this) {
-                currentError = null;
-                FTLController.this.moduleContextResponse = moduleContextResponse;
-                if (waiters) {
-                    this.notifyAll();
-                    waiters = false;
-                }
-            }
+    final StreamObserver<ModuleContextResponse> moduleObserver = new ModuleObserver();
 
-        }
+    private static volatile FTLController controller;
 
-        @Override
-        public void onError(Throwable throwable) {
-            log.error("GRPC connection error", throwable);
-            synchronized (this) {
-                currentError = throwable;
-                if (waiters) {
-                    this.notifyAll();
-                    waiters = false;
+    /**
+     * TODO: look at how init should work, this is terrible and will break dev mode
+     */
+    public static FTLController instance() {
+        if (controller == null) {
+            synchronized (FTLController.class) {
+                if (controller == null) {
+                    controller = new FTLController();
                 }
             }
-            if (!closed) {
-                verbService.getModuleContext(ModuleContextRequest.newBuilder().setModule(moduleName).build(), moduleObserver);
-            }
-        }
-
-        @Override
-        public void onCompleted() {
-            onError(new RuntimeException("connection closed"));
         }
-    };
-
-    @PreDestroy
-    void shutdown() {
-
+        return controller;
     }
 
-    public FTLController(@ConfigProperty(name = "ftl.endpoint", defaultValue = "http://localhost:8892") URI uri,
-            @ConfigProperty(name = "ftl.module.name") String moduleName) {
-        this.moduleName = moduleName;
+    FTLController() {
+        String endpoint = System.getenv("FTL_ENDPOINT");
```

Comment:
Looks like we lost the default here?

Reply:
I'm not sure actually, but it is currently the desired protocol.

Reply:
I have added it back.